service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/gcp/ainotebook/post-startup.sh

```diff
@@ -47,6 +49,30 @@ function get_metadata_value() {
     "http://metadata/computeMetadata/v1/$1"
 }
 
+#######################################
+# Set guest attributes on GCE. Used here to log completion status of the script.
+# See https://cloud.google.com/compute/docs/metadata/manage-guest-attributes
+# Arguments:
+#   $1: The guest attribute domain and key IE startup_script/status
+#   $2  The data to write to the guest attribute
+#######################################
+function set_guest_attributes() {
+  curl -s -X PUT --data "$2" \
+    -H "Metadata-Flavor: Google" \
+    "http://metadata.google.internal/computeMetadata/v1/instance/guest-attributes/$1"
+}
+
+# Write error status and message to guest attributes should an error occur.
+function error_handler {
+  set_guest_attributes "$STATUS_ATTRIBUTE" "ERROR"
+  set_guest_attributes "$MESSAGE_ATTRIBUTE" "Error on line $1, command \"$2\". See $OUTPUT_FILE for more information"
+  exit 1
+}
+trap 'error_handler $LINENO $BASH_COMMAND' ERR
```

Comment:
Reflecting a bit further, I'm unclear what trips the ERR trap specifically.
I have generally preferred trapping EXIT and then just having an if/then/else flow based on the exit code.

```
function exit_handler {
  local exit_code="$1"
  local line_no="$2"
  local command="$3"

  # Success! Set the guest attributes and exit cleanly
  if [[ "${exit_code}" -eq 0 ]]; then
    set_guest_attributes "$STATUS_ATTRIBUTE" "COMPLETE"
    exit 0
  fi

  # Write error status and message to guest attributes
  set_guest_attributes "$STATUS_ATTRIBUTE" "ERROR"
  set_guest_attributes "$MESSAGE_ATTRIBUTE" "Error on line $1, command \"$2\". See $OUTPUT_FILE for more information"
  exit "${exit_code}"
}
trap 'exit_handler $? $LINENO $BASH_COMMAND' EXIT
```

Reply:
This article captures (traps?) the essence and intention of the above approach:

https://medium.com/@dirk.avery/the-bash-trap-trap-ce6083f36700

Reply:
Thanks for the learning! Made the changes